sympy/solvers/solveset.py

```diff
@@ -2349,17 +2341,35 @@ def solvify(f, symbol, domain):
     >>> solvify(tan(x), x, S.Reals)
     [0]
     >>> solvify(exp(x) - 1, x, S.Complexes)
-
+    [2*_n*I*pi]
```

Comment:
Is the agreement that we can't start returning parametric solutions? If so then let's return 2 periods worth of values. Otherwise, I think that a parametric solution with the assumptions put on the dummy symbol is a good idea because each "single" parametric solution will be able to be substituted into the equation and demonstrate that it is a solution.

Reply:
`solvify` would then be an intermediate between `solve` and `solveset`.

Reply:
I thought that the whole point of `solvify` was to eliminate uses of `solve` by being an equivalent API but based over the top of `solveset`.

Reply:
So the question is whether it would be advantageous to return a parametric solution from solve that allows straightforward verification that it is a solution or a specific solution which doesn't allow one to know that this is one of many. Perhaps a flag `parametric` could be added to each to allow one to select between the two?

Reply:
Then solveset and solvify are different in that one gives the explicit set whereas the other embeds the set information on the symbol. This cannot be done for all sets (like unions and intervals that contain positive and negative values). Not sure what to do in those cases...

Reply:
We could make a symbol that has a set attached like `ElementOf('x', Integers)`.

A proper API for doing this would pass that information separately though.

Reply:
> solvify would then be an intermediate between solve and solveset.
> 
> I thought that the whole point of solvify was to eliminate uses of solve by being an equivalent API but based over the top of solveset.

I think @smichr given an appropriate justification. We can have a parametric solution for `solvify` that will be considered intermediate between `solve` and `solveset` in fact better than API than `solve`.

> This cannot be done for all sets (like unions and intervals that contain positive and negative values). Not sure what to do in those cases...

I think it also gives a solution for intervals. like in e.g. below `p1` & `p2`

https://github.com/sympy/sympy/blob/4766b20cdbb2608099459887eaa7b192f5433e2a/sympy/solvers/tests/test_solveset.py#L1325-L1333

Reply:
> I think @smichr given an appropriate justification

I don't consider that a justification. One of the problems we have in sympy generally is that the precise intended meanings of many functions is not defined in any formal sense. In this sense `solveset` is much better than `solve` but it is still not fully well defined.

The docstring for `solvify` says:
```
Solves an equation using solveset and returns the solution in accordance
    with the `solve` output API.
```
My understanding is that this was intended to allow a call to `solve` to be replaced directly with a call to `solvify`. This is therefore a strategy for migrating usage of `solve` within the sympy codebase to using `solveset` indirectly.

If that is the intended purpose of `solvify` then it's output should match `solve` exactly.

If there is some other intended use for `solvify` then I think it's a waste of time and we shouldn't work on it at all. We already have too many solve APIs. If we want to make a new one then we need to have a really good think about why we are doing that.

Reply:
I think the output API of `solve` in the sense of returning a list of solutions (instead of set in case of `solveset`), isn't it? 

IMHO, that doesn't mean it should give exactly the same solution as that of `solve`. It may also give a better solution when compared to `solve`.

If this doesn't make sense it's better to replace a solution with `n` (parametric) with a solution considering periodicity (as shown here https://github.com/sympy/sympy/pull/21174#issuecomment-819781275).

Reply:
If we wanted to design an API that was better than `solve` then we can make something much better than `solvify`. If we really want to support these kinds of parametrised solutions then it would be better to return the symbols and sets like:
```python
>>> newsolve(sin(x), x, Reals)
[(n*pi, [n], [Integers], True)]
>>> newsolve(x, x, Reals)
[(0, [], [], True)]
>>> newsolve(a*x, x, Reals)
[(0, [], [], a != 0),
 (y, [y], [Reals], a = 0)]
>>> newsolve(a*x, x, Reals, generic=True)
[(0, [], [], a != 0)]
```
This is essentially a union of `ImageSet` like returned by `solveset` except that we return the symbols and sets explicitly. This has all of the advantages of `solveset` like being able to express the same sets of solutions except that the result is structured and directly usable whereas `solveset` is practically impossible to use because you get a mangled combination of Complement, Union, Intersection etc.

I think that trying to improve `solvify` to be something more than `solve` is a mistake and a waste of time. If the plan is for `solveset` to actually replace `solve` (at least internally) then making `solvify` as feature complete as `solve` is worth doing but it needs to have exactly the same behaviour.

Reply:
> whereas `solveset` is practically impossible to use because you get a mangled combination of Complement, Union, Intersection etc.

The same solutions that are mangled in terms of `solveset` will be complicated no matter what form you return them in, won't they? As an exercise it would be worth looking at the types of solutions that `sovleset` gives to see if a better format (like you have shown) can be applied consistently for them.

Reply:
> The same solutions that are mangled in terms of `solveset` will be complicated no matter what form you return them in, won't they?

Not necessarily. The problem with `solveset` is that it's hard to get the solutions in the form of expressions because they are buried in an abritrarily complex arrangement of sets but as far as I can tell all the complements and intersections are really just trying to represent boolean conditions. For example:
```python
In [2]: solveset(x**2 - y, x, Reals)
Out[2]: ℝ ∩ {-√y, √y}
```
That would just be:
```python
>>> newsolve(x**2 - y, x, Reals)
[(√y, [], [], y in ℝ)
 (-√y, [], [], y in ℝ)]
```